### ISSUE TITLE
Scrollback attempting to access index out of bounds

### ISSUE THREAD

**Issue by matthiaskrgr** [jwilm/alacritty#1400]
rustc 1.28.0-nightly (e3bf634e0 2018-06-28)
Linux t470 4.17.0-2-MANJARO #1 SMP PREEMPT Fri Jun 8 07:13:17 UTC 2018 x86_64 GNU/Linux
alacritty @ scrollback / bfd62ef

from time to time I get random crashes on scrollback when there is a lot of text put out very fast.

I managed to get a backtrace by running cppcheck --enable=all --inconclusive --debug --verbose -j 4 on the llvm repo which made alacritty crash one time:

````
thread 'main' panicked at 'assertion failed: requested < self.inner.len()', src/grid/storage.rs:201:9
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:511
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:445
   6: <alacritty::grid::storage::Storage<T>>::compute_index
             at src/grid/storage.rs:201
   7: <alacritty::grid::storage::Storage<T> as core::ops::index::Index<usize>>::index
             at src/grid/storage.rs:285
   8: <alacritty::grid::Grid<T> as core::ops::index::Index<usize>>::index
             at src/grid/mod.rs:512
   9: <alacritty::grid::GridIterator<'a, T> as core::iter::iterator::Iterator>::next
             at src/grid/mod.rs:471
  10: <alacritty::term::Term as alacritty::selection::SemanticSearch>::semantic_search_right
             at src/term/mod.rs:63
  11: alacritty::selection::Selection::span_semantic
             at src/selection.rs:181
  12: alacritty::selection::Selection::to_span
             at src/selection.rs:158
  13: alacritty::term::Term::selection_to_string
             at src/term/mod.rs:980
  14: <alacritty::event::ActionContext<'a, N> as alacritty::input::ActionContext>::copy_selection
             at ./src/event.rs:62
  15: <alacritty::input::Processor<'a, A>>::on_mouse_release
             at ./src/input.rs:454
  16: <alacritty::input::Processor<'a, A>>::mouse_input
             at ./src/input.rs:559
  17: <alacritty::event::Processor<N>>::handle_event
             at ./src/event.rs:312
  18: <alacritty::event::Processor<N>>::process_events::{{closure}}
             at ./src/event.rs:421
  19: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &'a mut F>::call_mut
             at /checkout/src/libcore/ops/function.rs:261
  20: winit::platform::platform::x11::EventsLoop::process_event
             at /home/matthias/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.15.0/src/platform/linux/x11/mod.rs:670
  21: winit::platform::platform::x11::EventsLoop::poll_events
             at /home/matthias/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.15.0/src/platform/linux/x11/mod.rs:212
  22: winit::platform::platform::EventsLoop::poll_events
             at /home/matthias/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.15.0/src/platform/linux/mod.rs:467
  23: winit::EventsLoop::poll_events
             at /home/matthias/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.15.0/src/lib.rs:225
  24: alacritty::window::Window::poll_events
             at ./src/window.rs:280
  25: <alacritty::event::Processor<N>>::process_events
             at ./src/event.rs:435
  26: alacritty::run
             at src/main.rs:179
  27: alacritty::main
             at src/main.rs:62
  28: std::rt::lang_start::{{closure}}
             at /checkout/src/libstd/rt.rs:74
  29: std::panicking::try::do_call
             at libstd/rt.rs:59
             at libstd/panicking.rs:310
  30: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:105
  31: std::rt::lang_start_internal
             at libstd/panicking.rs:289
             at libstd/panic.rs:392
             at libstd/rt.rs:58
  32: std::rt::lang_start
             at /checkout/src/libstd/rt.rs:74
  33: main
  34: __libc_start_main
  35: _start
````

**Comment by chrisduerr** [comment]
So thanks to the help of @matthiaskrgr, we were able to track down the cause of this issue, or at least a way to reproduce it. In the process of doing so we've also noticed another related crash.

Here's a quick guide to reproduce these issues:

Semantic Crash:
 - Run `yes`
 - Start semantic selection with a doubleclick
 - Crash

Scrolling Crash:
 - Run `yes`
 - Try scrolling while `yes` is running

These issues both are related to the way the scrollback buffer works, with the semantic selection it will attempt to start a selection from the start point, however this is long out of range already, since `yes` pushed the starting point out of bounds.

The scrolling is probably a related issue, I haven't investigated this yet but I'd assume that since we're scrolling up the buffer now has an offset. When we run `yes`, the buffer doesn't automatically scroll down, but instead stays at the position where it initially was when scrolling. Then when the visible area is pushed out of bounds, a crash occurs.
